### DISCUSSION TITLE
Unable to use pycaret with numpy array

### DISCUSSION THREAD

**Discussion by binary-alkemi** [pycaret/pycaret#3352] [Q&A]
Hello Community members,

I am building a classification model with pycaret with the command:

`exp_class101 = setup(data=data , target = 'pubchem_activity' , session_id=123, fold_shuffle=True) `

However, I am getting error as follows:

`TypeError: Cannot compare types 'ndarray(dtype=object)' and 'float'`

My data looks like this:
```
,Morgan2FP,pubchem_activity
0,[0. 0. 1. ... 0. 0. 0.],0
1,[0. 0. 0. ... 0. 0. 0.],0
2,[0. 0. 0. ... 0. 0. 0.],0
3,[0. 0. 0. ... 0. 0. 0.],0
4,[0. 1. 0. ... 0. 0. 0.],1
5,[0. 1. 0. ... 0. 0. 0.],1
```

Where the target variable is PubChem activity. 

Can anyone suggest using the NumPy array as an input to pycaret classification?

**Comment by tvdboom** [comment]
Hi @binary-alkemi, I don't understand the format of your data. Pycaret accepts a numpy array of the form (n_samples, n_columns). You have something completely different. Given that you have indices and column names, you should make a pd.DataFrame from your data and provide that to pycaret as numpy array. If you really want the np.array (in which case you are gonna lose the column names), you should convert your data to a 2d array where the last column is your target column, and remove the indices.

**Comment by binary-alkemi** [comment]
Hi @tvdboom ,

Thanks for your reply. I understand converting things to Pandas Dataframe. But, do you think is it correct way of doing it? As each X-value np.array is a fingerprint, chemical structure encoded in numbers.

Each X-value is numpy array of 0's and 1's of length 2048.  I have such ~ 3000 rows. 

Are you suggesting to use each array element as column , so that I will have 2048 columns and 3000 rows? I tried this earlier and it works, but I am not sure if this is a right approach.

I asked this question because I can use same input format for RandomForestClassifier, but not in Pycaret.

**Comment by tvdboom** [comment]
Pycaret always uses pd.DataFrame as data structure. If you provide a np. array in `setup`, the type is converted internally.  So yes, I would say that's the right approach. Having 2048 columns should be no issue. Different models accept different input formats and we had to choose one that works for all, as well as for the preprocessing pipeline, plots, etc... hence we chose dataframe as the de facto format.

Ideally, you would use a dataframe with sparse columns, since your data is binary. But Pycaret will convert it internally to the full matrix so that won't help. Proper support for sparse datasets is planned for a future release.

**Reply by BradKML** [reply]
Need an update on this since there are times when people have columns with a lot of categories (e.g. consumer products), and that the idea of a "sparse column" would help with memory usage.
